external-crates/move/crates/move-analyzer/editors/code/src/configuration.ts

```diff
@@ -26,29 +45,96 @@ export class Configuration {
-        if (serverPath.length === 0) {
-            // The default value of the `server.path` setting is 'move-analyzer'.
-            // A user may have over-written this default with an empty string value, ''.
-            // An empty string cannot be an executable name, so instead use the default.
-            return defaultName;
+        const serverPath = this.configuration.get<string | null>('server.path') ?? this.defaultServerPath.fsPath;
+        if (serverPath.startsWith('~/')) {
+            return os.homedir() + serverPath.slice('~'.length);
+        }
+        return Path.resolve(serverPath);
+    }
+
+    /**
+     * Installs language server binary in the default location unless a user-specified
+     * (but not default) server location already contains a server binary.
+     *
+     * @returns `true` if server binary installation succeeded, `fase` otherwise.
+     */
+    async installServerBinary(extensionContext: vscode.ExtensionContext): Promise<boolean> {
+        log.info('Installing language server binary');
+
+        // Check if server binary is bundled with the extension
+        const bundledServerPath = vscode.Uri.joinPath(extensionContext.extensionUri,
+                                                    'language-server',
+                                                    this.serverName);
+        const bundledServerExists = await vscode.workspace.fs.stat(bundledServerPath).then(
+            () => true,
+            () => false,
+        );
+
+        const serverPath = vscode.Uri.file(this.serverPath);
+        const serverPathExists = await vscode.workspace.fs.stat(serverPath).then(
+            () => true,
+            () => false,
+        );
+
+        if (serverPath.fsPath !== this.defaultServerPath.fsPath) {
+            // User has overwritten default server path (need to compare paths as comparing URIs fails
+            // for some reason even if one is initialized from another).
+            if (!serverPathExists) {
+                // The server binary on user-overwritten path does not exist so warn the user about it.
+                // Need to use modal messages, otherwise the promise returned from is not resolved
+                // (and the extension blocks), which can confusing.
+                const items: vscode.MessageItem = { title: 'OK', isCloseAffordance: true };
+                await vscode.window.showInformationMessage(
+                    `The move-analyzer binary at the user-specified path ('${this.serverPath}') ` +
+                    'is not available. Put the binary in the path or reset user settings and ' +
+                    'reinstall the extension to use the bundled binary.',
+                    { modal: true },
+                    items,
+                );
+                return Promise.resolve(false);
+            } // Otherwise simply return and use the existing user-specified binary.
+            return Promise.resolve(true);
```

Comment:
I think it should be possible to return just `return false` / `return true` here and elsewhere, without `Promise.resolve` (unless some version of ts stops that from being possible).

Reply:
Quite true! I am only starting with TS so thank you for the hint and please keep them coming!